File: Packages/TabCore/Sources/TabCore/SplitwiseImport.swift
import Foundation

/// Parses a Splitwise group CSV export into trip-shaped rows.
///
/// Splitwise exports one signed **net** column per person (`paid - owed`) rather
/// than the raw payer/split structure. Because `BalanceEngine` derives all
/// balances from per-person nets, any reconstruction that preserves each
/// person's net is balance-exact — even where the original split shape can't be
/// recovered. This parser reconstructs payments + splits that (a) sum exactly to
/// the row's `Cost` (the DB and `SplitCalculator` require this) and (b) preserve
/// each person's net, using deterministic minor-unit remainder distribution.
///
/// The parser is name-based and UUID-free: rows reference people by the display
/// name in the header. The app layer maps names onto trip-person UUIDs. All
/// money math is done in integer minor units (never `Double`).
public enum SplitwiseImport {

    // MARK: - Output shapes

    /// One person's amount within a row. `amount` is always positive.
    public struct Share: Equatable, Sendable {
        public let person: String
        public let amount: Decimal

        public init(person: String, amount: Decimal) {
            self.person = person
            self.amount = amount
        }
    }

    /// A real expense. `payments` (who paid) and `splits` (who owes) each sum to
    /// `total`. People with a zero net are omitted from both.
    public struct ExpenseRow: Equatable, Sendable {
        public let date: Date
        public let description: String
        public let category: String
        public let currency: String
        public let total: Decimal
        public let payments: [Share]
        public let splits: [Share]

        public init(
            date: Date,
            description: String,
            category: String,
            currency: String,
            total: Decimal,
            payments: [Share],
            splits: [Share]
        ) {
            self.date = date
            self.description = description
            self.category = category
            self.currency = currency
            self.total = total
            self.payments = payments
            self.splits = splits
        }
    }

    /// A money transfer between two people. `from` paid `to` (matching tab's
    /// `Settlement.fromUserID -> toUserID` convention).
    public struct SettlementRow: Equatable, Sendable {
        public let date: Date
        public let description: String
        public let currency: String
        public let amount: Decimal
        public let from: String
        public let to: String

        public init(
            date: Date,
            description: String,
            currency: String,
            amount: Decimal,
            from: String,
            to: String
        ) {
            self.date = date
            self.description = description
            self.currency = currency
            self.amount = amount
            self.from = from
            self.to = to
        }
    }

    public enum Row: Equatable, Sendable {
        case expense(ExpenseRow)
        case settlement(SettlementRow)
    }

    /// A non-fatal issue with a single source line. The row may have been skipped
    /// or imported with an adjustment; `message` says which.
    public struct Warning: Equatable, Sendable {
        public let line: Int
        public let message: String

        public init(line: Int, message: String) {
            self.line = line
            self.message = message
        }
    }

    public struct Result: Equatable, Sendable {
        public let people: [String]
        public let rows: [Row]
        public let warnings: [Warning]

        public init(people: [String], rows: [Row], warnings: [Warning]) {
            self.people = people
            self.rows = rows
            self.warnings = warnings
        }

        public var expenses: [ExpenseRow] {
            rows.compactMap { if case let .expense(e) = $0 { return e } else { return nil } }
        }

        public var settlements: [SettlementRow] {
            rows.compactMap { if case let .settlement(s) = $0 { return s } else { return nil } }
        }
    }

    public enum ParseError: Error, Equatable, Sendable {
        case empty
        case missingHeader
        /// The five leading columns weren't the expected Splitwise header.
        case unexpectedHeader(found: [String])
        case noPeople
    }

    static let expectedLeadingColumns = ["Date", "Description", "Category", "Cost", "Currency"]
    static let settlementCategory = "Payment"
    static let settleAllDescription = "Settle all balances"
    static let totalBalanceDescription = "Total balance"

    // MARK: - Entry point

    public static func parse(_ text: String) throws -> Result {
        // Strip a leading UTF-8 BOM so the parser is self-contained regardless of
        // how the file was read.
        let input = text.hasPrefix("\u{FEFF}") ? String(text.dropFirst()) : text

        let records = splitRecords(input)
        guard let header = records.first else { throw ParseError.empty }

        let leading = header.fields.prefix(expectedLeadingColumns.count).map(trimmed)
        guard leading.count == expectedLeadingColumns.count else { throw ParseError.missingHeader }
        for (found, expected) in zip(leading, expectedLeadingColumns)
        where found.caseInsensitiveCompare(expected) != .orderedSame {
            throw ParseError.unexpectedHeader(found: leading)
        }

        // Raw header columns keep their positions (for reading cells); `people`
        // is the de-duplicated, first-appearance-ordered list that callers use.
        let columnNames = header.fields.dropFirst(expectedLeadingColumns.count)
            .map(trimmed)
            .filter { !$0.isEmpty }
        var people: [String] = []
        var seen = Set<String>()
        for name in columnNames where seen.insert(name).inserted { people.append(name) }
        guard !people.isEmpty else { throw ParseError.noPeople }

        var rows: [Row] = []
        var warnings: [Warning] = []

        if people.count != columnNames.count {
            warnings.append(Warning(
                line: header.line,
                message: "Duplicate person names in the header are merged into one person."
            ))
        }

        for record in records.dropFirst() {
            parseRow(record, columnNames: columnNames, people: people, into: &rows, warnings: &warnings)
        }

        return Result(people: people, rows: rows, warnings: warnings)
    }

    // MARK: - Row parsing

    private static func parseRow(
        _ record: (line: Int, fields: [String]),
        columnNames: [String],
        people: [String],
        into rows: inout [Row],
        warnings: inout [Warning]
    ) {
        let fields = record.fields
        // Blank separator line.
        if fields.allSatisfy({ trimmed($0).isEmpty }) { return }

        let dateString = field(fields, 0)
        let description = field(fields, 1)
        let category = field(fields, 2)
        let costString = field(fields, 3)
        let currency = CurrencyCatalog.normalizedCode(field(fields, 4))

        // The export's trailing summary row.
        if description.caseInsensitiveCompare(totalBalanceDescription) == .orderedSame { return }

        guard let date = parseDate(dateString) else {
            warnings.append(Warning(line: record.line, message: "Unreadable date \"\(dateString)\"; row skipped."))
            return
        }
        guard let total = parseDecimal(costString), total > 0 else {
            warnings.append(Warning(line: record.line, message: "Unreadable or zero amount; row skipped."))
            return
        }

        // Per-person nets, summing any duplicate-named columns together. A
        // non-empty cell that won't parse skips the whole row rather than
        // silently treating it as zero.
        var netByName: [String: Decimal] = [:]
        for (index, name) in columnNames.enumerated() {
            let raw = field(fields, expectedLeadingColumns.count + index)
            if raw.isEmpty { continue }
            guard let value = parseDecimal(raw) else {
                warnings.append(Warning(line: record.line, message: "Unreadable amount \"\(raw)\"; row skipped."))
                return
            }
            netByName[name, default: 0] += value
        }
        let nets: [(person: String, net: Decimal)] = people.compactMap { name in
            let net = netByName[name] ?? 0
            return net != 0 ? (name, net) : nil
        }

        let isSettlement =
            category.caseInsensitiveCompare(settlementCategory) == .orderedSame
            || description.caseInsensitiveCompare(settleAllDescription) == .orderedSame

        if isSettlement {
            let positives = nets.filter { $0.net > 0 }
            let negatives = nets.filter { $0.net < 0 }
            if positives.count == 1, negatives.count == 1 {
                rows.append(.settlement(SettlementRow(
                    date: date,
                    description: description,
                    currency: currency,
                    amount: total,
                    from: positives[0].person,
                    to: negatives[0].person
                )))
                return
            }
            // A "payment"-shaped row that isn't a clean two-party transfer: fall
            // through to expense reconstruction (balance-identical) with a note.
            warnings.append(Warning(
                line: record.line,
                message: "Payment row \"\(description)\" wasn't a two-person transfer; imported as an expense."
            ))
        }

        guard let reconstructed = reconstruct(
            total: total,
            currency: currency,
            nets: nets,
            people: people,
            line: record.line,
            warnings: &warnings
        ) else { return }

        rows.append(.expense(ExpenseRow(
            date: date,
            description: description,
            category: category,
            currency: currency,
            total: total,
            payments: reconstructed.payments,
            splits: reconstructed.splits
        )))
    }

    // MARK: - Net -> payments/splits reconstruction

    /// Rebuilds payments + splits from per-person nets, anchored on `total`.
    ///
    /// Debtors (net < 0) owe `-net` and paid nothing. Payers (net > 0) share the
    /// remaining `total - debtorOwed` as their owed amount; each payer's paid is
    /// `net + owed`. All arithmetic is in the currency's integer minor units so
    /// sums are exact; the remainder goes to the largest fractional share first
    /// (lowest name on ties), mirroring `BalanceEngine.distributePairs`. Returns
    /// nil (row skipped) for data that can't be reconstructed into a valid,
    /// total-summing expense.
    static func reconstruct(
        total: Decimal,
        currency: String,
        nets: [(person: String, net: Decimal)],
        people: [String],
        line: Int,
        warnings: inout [Warning]
    ) -> (payments: [Share], splits: [Share])? {
        let multiplier = CurrencyCatalog.minorUnitMultiplier(for: currency)
        let costMinor = minorUnits(total, multiplier)

        let netMinor = nets
            .map { (person: $0.person, net: minorUnits($0.net, multiplier)) }
            .filter { $0.net != 0 }

        let debtors = netMinor.filter { $0.net < 0 }
        let payers = netMinor.filter { $0.net > 0 }

        var owedMinor: [(person: String, amount: Int)] = []
        var paidMinor: [(person: String, amount: Int)] = []

        guard !payers.isEmpty else {
            if debtors.isEmpty {
                // Every net is zero: a self-contained expense. Attribute it to the
                // first listed person so it stays a valid, balance-neutral row.
                guard let solo = people.first else { return nil }
                warnings.append(Warning(
                    line: line,
                    message: "Row has no balances; recorded as paid in full by \(solo)."
                ))
                return ([Share(person: solo, amount: total)], [Share(person: solo, amount: total)])
            }
            warnings.append(Warning(line: line, message: "Row doesn't balance (no payer); skipped."))
            return nil
        }

        let debtorOwedTotal = debtors.reduce(0) { $0 - $1.net }   // sum of -net
        let payersOwedTotal = costMinor - debtorOwedTotal
        guard payersOwedTotal >= 0 else {
            warnings.append(Warning(line: line, message: "Row doesn't balance (debts exceed the amount); skipped."))
            return nil
        }

        for debtor in debtors {
            owedMinor.append((debtor.person, -debtor.net))
        }

        // Distribute the payers' shared owed amount proportionally to surplus.
        // The proportion is computed in Decimal (38-digit precision) to avoid
        // Int64 overflow on large expenses, then floored to integer minor units;
        // the remainder goes to the largest fractional share first (lowest name
        // on ties), mirroring BalanceEngine.distributePairs.
        let totalSurplus = payers.reduce(0) { $0 + $1.net }
        let owedTotalDecimal = Decimal(payersOwedTotal)
        let surplusDecimal = Decimal(totalSurplus)
        let sortedPayers = payers.sorted { $0.person < $1.person }
        var shares = [Int](repeating: 0, count: sortedPayers.count)
        var fractions: [(index: Int, fraction: Decimal)] = []
        var allocated = 0
        for (index, payer) in sortedPayers.enumerated() {
            let raw = owedTotalDecimal * Decimal(payer.net) / surplusDecimal
            let floored = floorToInt(raw)
            shares[index] = floored
            allocated += floored
            fractions.append((index, raw - Decimal(floored)))
        }
        var leftover = payersOwedTotal - allocated
        let order = fractions.sorted {
            $0.fraction != $1.fraction ? $0.fraction > $1.fraction
                : sortedPayers[$0.index].person < sortedPayers[$1.index].person
        }
        var next = 0
        while leftover > 0 {
            shares[order[next % order.count].index] += 1
            leftover -= 1
            next += 1
        }

        for (index, payer) in sortedPayers.enumerated() {
            let owed = shares[index]
            if owed > 0 { owedMinor.append((payer.person, owed)) }
            paidMinor.append((payer.person, payer.net + owed))
        }

        // Absorb any net rounding drift (cents that don't sum to zero in the
        // source) into the largest payer so payments total the cost exactly.
        let paidSum = paidMinor.reduce(0) { $0 + $1.amount }
        let drift = costMinor - paidSum
        if drift != 0, let largest = paidMinor.indices.max(by: { paidMinor[$0].amount < paidMinor[$1].amount }) {
            paidMinor[largest].amount += drift
            if abs(drift) > 2 {
                warnings.append(Warning(
                    line: line,
                    message: "Row's amounts didn't sum cleanly; adjusted by \(decimalString(drift, multiplier)) to balance."
                ))
            }
        }

        // Final guard: never emit an expense whose payments or splits don't sum
        // exactly to the total, or that has a negative share (the DB rejects it).
        let finalPaid = paidMinor.reduce(0) { $0 + $1.amount }
        let finalOwed = owedMinor.reduce(0) { $0 + $1.amount }
        guard finalPaid == costMinor, finalOwed == costMinor,
              paidMinor.allSatisfy({ $0.amount >= 0 }), owedMinor.allSatisfy({ $0.amount >= 0 })
        else {
            warnings.append(Warning(line: line, message: "Row's amounts don't balance; skipped."))
            return nil
        }

        let payments = paidMinor
            .filter { $0.amount > 0 }
            .map { Share(person: $0.person, amount: fromMinor($0.amount, multiplier)) }
        let splits = owedMinor
            .filter { $0.amount > 0 }
            .map { Share(person: $0.person, amount: fromMinor($0.amount, multiplier)) }
        return (payments, splits)
    }

    // MARK: - Primitive parsing

    /// Splits CSV text into records, preserving 1-based source line numbers.
    /// Handles `\r\n`, `\r`, and `\n`, and quoted fields containing commas
    /// (RFC 4180, single-line fields — Splitwise never embeds newlines in a
    /// field). Fully blank lines are kept here (callers skip them) so warnings
    /// can reference real line numbers.
    static func splitRecords(_ text: String) -> [(line: Int, fields: [String])] {
        var records: [(line: Int, fields: [String])] = []
        let normalized = text.replacingOccurrences(of: "\r\n", with: "\n")
            .replacingOccurrences(of: "\r", with: "\n")
        for (offset, rawLine) in normalized.components(separatedBy: "\n").enumerated() {
            if rawLine.isEmpty { continue }
            records.append((line: offset + 1, fields: parseFields(rawLine)))
        }
        return records
    }

    /// RFC 4180 field tokenizer for a single line.
    static func parseFields(_ line: String) -> [String] {
        var fields: [String] = []
        var current = ""
        var inQuotes = false
        var iterator = line.makeIterator()
        var pending: Character? = iterator.next()
        while let char = pending {
            pending = iterator.next()
            if inQuotes {
                if char == "\"" {
                    if pending == "\"" {            // escaped quote
                        current.append("\"")
                        pending = iterator.next()
                    } else {
                        inQuotes = false
                    }
                } else {
                    current.append(char)
                }
            } else {
                switch char {
                case "\"": inQuotes = true
                case ",":
                    fields.append(current)
                    current = ""
                default:
                    current.append(char)
                }
            }
        }
        fields.append(current)
        return fields
    }

    /// Parses a plain signed decimal (`-12.34`). Rejects anything with grouping
    /// separators or stray characters rather than silently truncating a
    /// partially-parsed token (e.g. `1,234.56` would otherwise become `1`).
    static func parseDecimal(_ string: String) -> Decimal? {
        let cleaned = trimmed(string)
        guard isPlainDecimal(cleaned) else { return nil }
        return Decimal(string: cleaned, locale: posix)
    }

    private static func isPlainDecimal(_ string: String) -> Bool {
        var characters = Substring(string)
        if let first = characters.first, first == "-" || first == "+" {
            characters = characters.dropFirst()
        }
        guard !characters.isEmpty else { return false }
        var sawDot = false
        var sawDigit = false
        for character in characters {
            if character == "." {
                if sawDot { return false }
                sawDot = true
            } else if character >= "0" && character <= "9" {
                sawDigit = true
            } else {
                return false
            }
        }
        return sawDigit
    }

    /// `yyyy-MM-dd` -> the calendar day anchored at 12:00 UTC, matching tab's
    /// `expense_date` storage convention (`ExpenseDates`).
    static func parseDate(_ string: String) -> Date? {
        let parts = trimmed(string).split(separator: "-")
        guard parts.count == 3,
              let year = Int(parts[0]), let month = Int(parts[1]), let day = Int(parts[2])
        else { return nil }
        var components = DateComponents()
        components.year = year
        components.month = month
        components.day = day
        components.hour = 12
        return utcCalendar.date(from: components)
    }

    // MARK: - Helpers

    private static func field(_ fields: [String], _ index: Int) -> String {
        index < fields.count ? trimmed(fields[index]) : ""
    }

    private static func trimmed(_ value: String) -> String {
        value.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func minorUnits(_ amount: Decimal, _ multiplier: Decimal) -> Int {
        var scaled = amount * multiplier
        var rounded = Decimal()
        NSDecimalRound(&rounded, &scaled, 0, .plain)
        return (rounded as NSDecimalNumber).intValue
    }

    private static func floorToInt(_ value: Decimal) -> Int {
        var input = value
        var rounded = Decimal()
        NSDecimalRound(&rounded, &input, 0, .down)
        return (rounded as NSDecimalNumber).intValue
    }

    private static func fromMinor(_ minor: Int, _ multiplier: Decimal) -> Decimal {
        Decimal(minor) / multiplier
    }

    private static func decimalString(_ minor: Int, _ multiplier: Decimal) -> String {
        "\(fromMinor(minor, multiplier))"
    }

    private static let posix = Locale(identifier: "en_US_POSIX")

    private static let utcCalendar: Calendar = {
        var calendar = Calendar(identifier: .iso8601)
        calendar.timeZone = TimeZone(identifier: "UTC")!
        return calendar
    }()
}

File: Apps/TabAndroid/core/domain/src/main/kotlin/com/rithwikshetty/tab/domain/SplitwiseImport.kt
package com.rithwikshetty.tab.domain

import java.math.BigDecimal
import java.math.BigInteger
import java.math.RoundingMode
import java.time.DateTimeException
import java.time.LocalDate
import java.time.ZoneOffset
import java.time.ZonedDateTime

public data class ImportShare(
    public val person: String,
    public val amount: BigDecimal,
)

public sealed interface SplitwiseRow {
    public data class Expense(
        public val date: ZonedDateTime,
        public val description: String,
        public val category: String,
        public val currency: String,
        public val total: BigDecimal,
        public val payments: List<ImportShare>,
        public val splits: List<ImportShare>,
    ) : SplitwiseRow

    public data class Settlement(
        public val date: ZonedDateTime,
        public val description: String,
        public val currency: String,
        public val amount: BigDecimal,
        public val from: String,
        public val to: String,
    ) : SplitwiseRow
}

public data class ImportWarning(
    public val line: Int,
    public val message: String,
)

public data class SplitwiseImportResult(
    public val people: List<String>,
    public val rows: List<SplitwiseRow>,
    public val warnings: List<ImportWarning>,
) {
    public val expenses: List<SplitwiseRow.Expense>
        get() = rows.filterIsInstance<SplitwiseRow.Expense>()
    public val settlements: List<SplitwiseRow.Settlement>
        get() = rows.filterIsInstance<SplitwiseRow.Settlement>()
}

public sealed class SplitwiseParseException(message: String) : IllegalArgumentException(message) {
    public data object Empty : SplitwiseParseException("The import is empty.")
    public data object MissingHeader : SplitwiseParseException("The Splitwise header is missing.")
    public data class UnexpectedHeader(public val found: List<String>) :
        SplitwiseParseException("Unexpected Splitwise header: $found")
    public data object NoPeople : SplitwiseParseException("The import contains no people.")
}

public object SplitwiseImport {
    private val leadingColumns = listOf("Date", "Description", "Category", "Cost", "Currency")
    private const val settlementCategory = "Payment"
    private const val settleAllDescription = "Settle all balances"
    private const val totalBalanceDescription = "Total balance"

    public fun parse(text: String): SplitwiseImportResult {
        val input = text.removePrefix("\uFEFF")
        val records = splitRecords(input)
        val header = records.firstOrNull() ?: throw SplitwiseParseException.Empty
        val leading = header.fields.take(leadingColumns.size).map(String::trim)
        if (leading.size != leadingColumns.size) throw SplitwiseParseException.MissingHeader
        if (leading.zip(leadingColumns).any { (found, expected) -> !found.equals(expected, true) }) {
            throw SplitwiseParseException.UnexpectedHeader(leading)
        }
        val columnNames = header.fields.drop(leadingColumns.size).map(String::trim).filter(String::isNotEmpty)
        val people = columnNames.distinct()
        if (people.isEmpty()) throw SplitwiseParseException.NoPeople
        val rows = mutableListOf<SplitwiseRow>()
        val warnings = mutableListOf<ImportWarning>()
        if (people.size != columnNames.size) {
            warnings += ImportWarning(header.line, "Duplicate person names in the header are merged into one person.")
        }
        records.drop(1).forEach { record ->
            parseRow(record, columnNames, people, rows, warnings)
        }
        return SplitwiseImportResult(people, rows, warnings)
    }

    private fun parseRow(
        record: CsvRecord,
        columnNames: List<String>,
        people: List<String>,
        rows: MutableList<SplitwiseRow>,
        warnings: MutableList<ImportWarning>,
    ) {
        if (record.fields.all { it.isBlank() }) return
        val dateText = record.field(0)
        val description = record.field(1)
        val category = record.field(2)
        val costText = record.field(3)
        val currency = CurrencyCatalog.normalizedCode(record.field(4))
        if (description.equals(totalBalanceDescription, true)) return
        val date = parseDate(dateText)
        if (date == null) {
            warnings += ImportWarning(record.line, "Unreadable date \"$dateText\"; row skipped.")
            return
        }
        val total = parseDecimal(costText)
        if (total == null || total <= BigDecimal.ZERO) {
            warnings += ImportWarning(record.line, "Unreadable or zero amount; row skipped.")
            return
        }
        val netByName = mutableMapOf<String, BigDecimal>()
        columnNames.forEachIndexed { index, name ->
            val raw = record.field(leadingColumns.size + index)
            if (raw.isEmpty()) return@forEachIndexed
            val value = parseDecimal(raw)
            if (value == null) {
                warnings += ImportWarning(record.line, "Unreadable amount \"$raw\"; row skipped.")
                return
            }
            netByName.merge(name, value, BigDecimal::add)
        }
        val nets = people.mapNotNull { person ->
            val net = netByName[person] ?: BigDecimal.ZERO
            if (net.compareTo(BigDecimal.ZERO) == 0) null else PersonNet(person, net)
        }
        val isSettlement = category.equals(settlementCategory, true) ||
            description.equals(settleAllDescription, true)
        if (isSettlement) {
            val positives = nets.filter { it.net > BigDecimal.ZERO }
            val negatives = nets.filter { it.net < BigDecimal.ZERO }
            if (positives.size == 1 && negatives.size == 1) {
                rows += SplitwiseRow.Settlement(
                    date,
                    description,
                    currency,
                    total,
                    positives.single().person,
                    negatives.single().person,
                )
                return
            }
            warnings += ImportWarning(
                record.line,
                "Payment row \"$description\" wasn't a two-person transfer; imported as an expense.",
            )
        }
        val reconstructed = reconstruct(total, currency, nets, people, record.line, warnings) ?: return
        rows += SplitwiseRow.Expense(
            date,
            description,
            category,
            currency,
            total,
            reconstructed.first,
            reconstructed.second,
        )
    }

    private fun reconstruct(
        total: BigDecimal,
        currency: String,
        nets: List<PersonNet>,
        people: List<String>,
        line: Int,
        warnings: MutableList<ImportWarning>,
    ): Pair<List<ImportShare>, List<ImportShare>>? {
        val multiplier = CurrencyCatalog.minorUnitMultiplier(currency)
        val costMinor = minorUnits(total, multiplier)
        val netMinor = nets.map { MinorNet(it.person, minorUnits(it.net, multiplier)) }
            .filter { it.net != BigInteger.ZERO }
        val debtors = netMinor.filter { it.net < BigInteger.ZERO }
        val payers = netMinor.filter { it.net > BigInteger.ZERO }
        val owed = mutableListOf<MinorShare>()
        val paid = mutableListOf<MinorShare>()
        if (payers.isEmpty()) {
            if (debtors.isEmpty()) {
                val solo = people.firstOrNull() ?: return null
                warnings += ImportWarning(line, "Row has no balances; recorded as paid in full by $solo.")
                return listOf(ImportShare(solo, total)) to listOf(ImportShare(solo, total))
            }
            warnings += ImportWarning(line, "Row doesn't balance (no payer); skipped.")
            return null
        }
        val debtorOwedTotal = debtors.fold(BigInteger.ZERO) { value, net -> value - net.net }
        val payersOwedTotal = costMinor - debtorOwedTotal
        if (payersOwedTotal < BigInteger.ZERO) {
            warnings += ImportWarning(line, "Row doesn't balance (debts exceed the amount); skipped.")
            return null
        }
        debtors.forEach { owed += MinorShare(it.person, it.net.negate()) }
        val sortedPayers = payers.sortedBy(MinorNet::person)
        val totalSurplus = sortedPayers.fold(BigInteger.ZERO) { value, payer -> value + payer.net }
        val floors = MutableList(sortedPayers.size) { BigInteger.ZERO }
        val fractions = mutableListOf<Pair<Int, BigDecimal>>()
        sortedPayers.forEachIndexed { index, payer ->
            val raw = BigDecimal(payersOwedTotal).multiply(BigDecimal(payer.net))
                .divide(BigDecimal(totalSurplus), 24, RoundingMode.HALF_UP)
            val floor = raw.setScale(0, RoundingMode.DOWN).toBigIntegerExact()
            floors[index] = floor
            fractions += index to raw.subtract(BigDecimal(floor))
        }
        var leftover = payersOwedTotal - floors.fold(BigInteger.ZERO, BigInteger::add)
        val order = fractions.sortedWith(
            compareByDescending<Pair<Int, BigDecimal>> { it.second }
                .thenBy { sortedPayers[it.first].person },
        )
        var next = 0
        while (leftover > BigInteger.ZERO) {
            val index = order[next % order.size].first
            floors[index] += BigInteger.ONE
            leftover -= BigInteger.ONE
            next++
        }
        sortedPayers.forEachIndexed { index, payer ->
            if (floors[index] > BigInteger.ZERO) owed += MinorShare(payer.person, floors[index])
            paid += MinorShare(payer.person, payer.net + floors[index])
        }
        val paidSum = paid.fold(BigInteger.ZERO) { value, share -> value + share.amount }
        val drift = costMinor - paidSum
        if (drift != BigInteger.ZERO) {
            val largestIndex = paid.indices.maxByOrNull { paid[it].amount }
            if (largestIndex != null) {
                paid[largestIndex] = paid[largestIndex].copy(amount = paid[largestIndex].amount + drift)
                if (drift.abs() > BigInteger.TWO) {
                    warnings += ImportWarning(
                        line,
                        "Row's amounts didn't sum cleanly; adjusted by ${fromMinor(drift, multiplier)} to balance.",
                    )
                }
            }
        }
        val valid = paid.fold(BigInteger.ZERO) { value, share -> value + share.amount } == costMinor &&
            owed.fold(BigInteger.ZERO) { value, share -> value + share.amount } == costMinor &&
            paid.all { it.amount >= BigInteger.ZERO } &&
            owed.all { it.amount >= BigInteger.ZERO }
        if (!valid) {
            warnings += ImportWarning(line, "Row's amounts don't balance; skipped.")
            return null
        }
        return paid.filter { it.amount > BigInteger.ZERO }
            .map { ImportShare(it.person, fromMinor(it.amount, multiplier)) } to
            owed.filter { it.amount > BigInteger.ZERO }
                .map { ImportShare(it.person, fromMinor(it.amount, multiplier)) }
    }

    internal fun splitRecords(text: String): List<CsvRecord> =
        text.replace("\r\n", "\n").replace('\r', '\n').split('\n').mapIndexedNotNull { index, line ->
            if (line.isEmpty()) null else CsvRecord(index + 1, parseFields(line))
        }

    internal fun parseFields(line: String): List<String> {
        val fields = mutableListOf<String>()
        val current = StringBuilder()
        var inQuotes = false
        var index = 0
        while (index < line.length) {
            val character = line[index]
            if (inQuotes) {
                if (character == '"' && index + 1 < line.length && line[index + 1] == '"') {
                    current.append('"')
                    index++
                } else if (character == '"') {
                    inQuotes = false
                } else {
                    current.append(character)
                }
            } else {
                when (character) {
                    '"' -> inQuotes = true
                    ',' -> {
                        fields += current.toString()
                        current.clear()
                    }
                    else -> current.append(character)
                }
            }
            index++
        }
        fields += current.toString()
        return fields
    }

    internal fun parseDecimal(value: String): BigDecimal? {
        val cleaned = value.trim()
        if (!Regex("[+-]?(?:\\d+(?:\\.\\d*)?|\\.\\d+)").matches(cleaned)) return null
        return cleaned.toBigDecimalOrNull()
    }

    internal fun parseDate(value: String): ZonedDateTime? = try {
        LocalDate.parse(value.trim()).atTime(12, 0).atZone(ZoneOffset.UTC)
    } catch (_: DateTimeException) {
        null
    }

    private fun minorUnits(amount: BigDecimal, multiplier: BigDecimal): BigInteger =
        amount.multiply(multiplier).setScale(0, RoundingMode.HALF_UP).toBigIntegerExact()

    private fun fromMinor(amount: BigInteger, multiplier: BigDecimal): BigDecimal =
        BigDecimal(amount).divide(multiplier)

    internal data class CsvRecord(val line: Int, val fields: List<String>) {
        fun field(index: Int): String = fields.getOrElse(index) { "" }.trim()
    }

    private data class PersonNet(val person: String, val net: BigDecimal)
    private data class MinorNet(val person: String, val net: BigInteger)
    private data class MinorShare(val person: String, val amount: BigInteger)
}
